plugins/inputs/clickhouse/clickhouse.go

```diff
@@ -0,0 +1,355 @@
+						Host:   net.JoinHostPort(c.Hostname, u.Port()),
+						User:   u.User,
+					}
+					connects = append(connects, c)
+				}
+			}
+		default:
+			connects = append(connects, connect{
+				url: u,
+			})
+		}
+	}
+
+	for _, conn := range connects {
+		if err := ch.tables(acc, &conn); err != nil {
+			acc.AddError(err)
+		}
+		for metric := range commonMetrics {
+			if err := ch.commonMetrics(acc, &conn, metric); err != nil {
+				acc.AddError(err)
+			}
+		}
+	}
+	return nil
+}
+
+func (ch *ClickHouse) clusterIncludeExcludeFilter() string {
+	if len(ch.ClusterInclude) == 0 && len(ch.ClusterExclude) == 0 {
+		return ""
+	}
+	var (
+		escape = func(in string) string {
+			return "'" + strings.NewReplacer(`\`, `\\`, `'`, `\'`).Replace(in) + "'"
+		}
+		makeFilter = func(expr string, args []string) string {
+			in := make([]string, 0, len(args))
+			for _, v := range args {
+				in = append(in, escape(v))
+			}
+			return "cluster " + expr + " (" + strings.Join(in, ", ") + ")"
+		}
+	)
+
+	if len(ch.ClusterInclude) != 0 {
+		return "WHERE " + makeFilter("IN", ch.ClusterInclude)
+	}
+
+	return "WHERE " + makeFilter("NOT IN", ch.ClusterExclude)
+}
+
+func (ch *ClickHouse) commonMetrics(acc telegraf.Accumulator, conn *connect, metric string) error {
+	var result []struct {
+		Metric string   `json:"metric"`
+		Value  chUInt64 `json:"value"`
+	}
+	if err := ch.execQuery(conn.url, commonMetrics[metric], &result); err != nil {
+		return err
+	}
+
+	tags := map[string]string{
+		"source": conn.Hostname,
+	}
+	if len(conn.Cluster) != 0 {
+		tags["cluster"] = conn.Cluster
+	}
+	if conn.ShardNum != 0 {
+		tags["shard_num"] = strconv.Itoa(conn.ShardNum)
+	}
+
+	fields := make(map[string]interface{})
+	for _, r := range result {
+		fields[internal.SnakeCase(r.Metric)] = uint64(r.Value)
+	}
+
+	acc.AddFields("clickhouse_"+metric, fields, tags)
+
+	return nil
+}
+
+func (ch *ClickHouse) tables(acc telegraf.Accumulator, conn *connect) error {
+	var parts []struct {
+		Database string   `json:"database"`
+		Table    string   `json:"table"`
+		Bytes    chUInt64 `json:"bytes"`
+		Parts    chUInt64 `json:"parts"`
+		Rows     chUInt64 `json:"rows"`
+	}
+
+	if err := ch.execQuery(conn.url, systemParts, &parts); err != nil {
+		return err
+	}
+	tags := map[string]string{
+		"source": conn.Hostname,
+	}
+	if len(conn.Cluster) != 0 {
+		tags["cluster"] = conn.Cluster
+	}
+	if conn.ShardNum != 0 {
+		tags["shard_num"] = strconv.Itoa(conn.ShardNum)
+	}
+	for _, part := range parts {
+		tags["table"] = part.Table
+		tags["database"] = part.Database
+		acc.AddFields("clickhouse_tables",
+			map[string]interface{}{
+				"bytes": uint64(part.Bytes),
+				"parts": uint64(part.Parts),
+				"rows":  uint64(part.Rows),
+			},
+			tags,
+		)
+	}
+	return nil
+}
+
+type clickhouseError struct {
+	StatusCode int
+	body       []byte
+}
+
+func (e *clickhouseError) Error() string {
+	return fmt.Sprintf("ClickHouse server returned an error code: %d\n%s", e.StatusCode, e.body)
```

Comment:
body could be quite large here. we probably want to display only the first 200 characters, escaped

Reply:
@ssoroka error message from Clickhouse could contain full stack trace, so 200 bytes is not enough

also, we don't expect errors on these queries, so full stack trace is really helpful for submit an issue to Clickhouse GitHub repo

Reply:
won't fix, this fix is necessary for complete review?

Reply:
We should limit this error.  It doesn't need to be by body length, we could for example take the first line or pull an error string from JSON, but any errors should all be single line and of reasonable length so that they can be included in a log message, and log lines should never span multiple lines.

It would be ideal if the error string matched the Go guidelines for error strings:
https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Telegraf will print the plugin name, so you can leave this part out of the error:
```
fmt.Sprintf("received error code %d: %v", code, err)
```